### PR TITLE
Vulkan: support all VkFormat for vertex and frame

### LINKED ISSUE #31: Vulkan: Support read/write of all image formats.

### PR DESCRIPTION
**Pull request by jaebaek**
Fixes #31

### CODE REVIEW

**Comment by jaebaek** [comment]
Split`Verifier::IsTexelEqualToExpected()` into tree parts:
* `std::vector<Value> GetActualValuesFromTexel()`
* `std::vector<Value> ScaleTexelValuesIfNeeded()`
* `Verifier::IsTexelEqualToExpected()`

PTAL

File: src/bit_copy.h [hunk 1]
```diff
@@ -45,15 +46,6 @@ class BitCopy {
...
-  // at |buffer| + |length_bytes|, this method ignores it. Note that
-  // this overflow is fine for BitCopy because we have only interest
-  // in data [0, length_bytes) bytes of |buffer|.
-  static void LeftShiftBufferBits(uint8_t* buffer,
-                                  uint8_t length_bytes,
-                                  uint8_t shift_bits);
-
   // Right-shift |length_bytes| bytes of buffer by |shift_bits| bits
   // based on assumption that |shift_bits| < 8. If there is underflow,
   // this method ignores it. Note that this underflow is fine for
```
Comment (dj2) [inline src/bit_copy.h]: RightShiftBufferBits is also removed, yea?

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -91,7 +92,7 @@ if (${AMBER_ENABLE_TESTS})
   )
 
   if (${Vulkan_FOUND})
```
Comment (dj2) [inline src/CMakeLists.txt]: This shouldn't be guarded by Vulkan anymore.

File: src/bit_copy.cc [hunk 3]
```diff
@@ -0,0 +1,230 @@
+// Copyright 2018 The Amber Authors.
```
Comment (dj2) [inline src/bit_copy.cc]: 2019

File: src/bit_copy.cc [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+  // and if |bits| is 64, |dst_bit_offset| must be 0. No component
+  // has |bits| bigger than 64.
+  data <<= dst_bit_offset;
+
+  CopyBits(dst, data, dst_bit_offset, bits);
+  return {};
+}
+
+// static
+void BitCopy::CopyMemoryToBuffer(uint8_t* dst,
```
Comment (dj2) [inline src/bit_copy.cc]: This is only used by verifyer, why not put it in the anonymous namespace of verifier.cc?

File: src/bit_copy.cc [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    ptr[i] = src[i];
+  }
+
+  data >>= src_bit_offset;
+
+  CopyBits(dst, data, 0, bits);
+}
+
+// static
+float BitCopy::HexFloatToFloat(const uint8_t* value, uint8_t bits) {
```
Comment (dj2) [inline src/bit_copy.cc]: ditto

File: src/bit_copy.cc [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+
+    bit_offset -= bit_offset;
+    bits -= target_bits;
+    ++dst;
+    ++ptr;
+  }
+}
+
+// static
+uint16_t BitCopy::FloatExponent(const uint32_t hex_float) {
```
Comment (dj2) [inline src/bit_copy.cc]: This is only used internally, move to anonymous namespace (same with other float methods below)

File: src/bit_copy.h [hunk 7]
```diff
@@ -12,33 +12,54 @@
...
 class BitCopy {
  public:
-  // Copy [0, bits) bits of src to [bit_offset, bit_offset + bits) of dst.
-  static void CopyValueToBuffer(uint8_t* dst,
-                                const Value& src,
-                                uint8_t bit_offset,
-                                uint8_t bits);
+  // Copy [0, bits) bits of |src| to
+  // [dst_bit_offset, dst_bit_offset + bits) of |dst|.
+  static Result CopyValueToBuffer(uint8_t* dst,
```
Comment (dj2) [inline src/bit_copy.h]: This is only used by vulkan_buffer, move to anoymous namespace

File: src/verifier.cc [hunk 8]
```diff
@@ -177,18 +188,204 @@ void SetupToleranceForTexels(const ProbeCommand* command,
...
+std::vector<Value> GetActualValuesFromTexel(const uint8_t* texel,
+                                            const Format* texel_format) {
+  assert(texel_format && !texel_format->GetComponents().empty());
+
+  std::vector<Value> actual_values(texel_format->GetComponents().size());
+  uint8_t bit_offset = 0;
+
+  for (size_t i = 0; i < texel_format->GetComponents().size(); ++i) {
+    const auto& component = texel_format->GetComponents()[i];
+    uint8_t actual[8] = {};
```
Comment (dj2) [inline src/verifier.cc]: This needs to be {0, 0, 0, 0, 0, 0, 0, 0} to make the cts happy

File: src/verifier.cc [hunk 9]
```diff
@@ -177,18 +188,204 @@ void SetupToleranceForTexels(const ProbeCommand* command,
...
+        float* ptr = reinterpret_cast<float*>(actual);
+        actual_values[i].SetDoubleValue(static_cast<double>(*ptr));
+      } else if (component.num_bits == 64) {
+        double* ptr = reinterpret_cast<double*>(actual);
+        actual_values[i].SetDoubleValue(*ptr);
+      } else {
+        assert(false && "Bits of component is not for double nor float type");
+      }
+    } else {
+      uint16_t* ptr16 = nullptr;
```
Comment (dj2) [inline src/verifier.cc]: Move the declaration down to the usages.

File: src/verifier.cc [hunk 10]
```diff
@@ -177,18 +188,204 @@ void SetupToleranceForTexels(const ProbeCommand* command,
...
+      uint32_t* ptr32 = nullptr;
+      uint64_t* ptr64 = nullptr;
+      switch (component.num_bits) {
+        case 8:
+          actual_values[i].SetDoubleValue(static_cast<double>(*actual));
```
Comment (dj2) [inline src/verifier.cc]: Why are these all set as doubles and not integers?

File: src/verifier.cc [hunk 11]
```diff
@@ -177,18 +188,204 @@ void SetupToleranceForTexels(const ProbeCommand* command,
...
+  return actual_values;
+}
+
+// If component mode of |texel_format| is FormatMode::kUNorm or
+// ::kSNorm or ::kSRGB, scale the corresponding value in |texel|.
+// Note that we do not scale values with FormatMode::kUInt, ::kSInt,
+// ::kUFloat, ::kSFloat.
+std::vector<Value> ScaleTexelValuesIfNeeded(const std::vector<Value>& texel,
+                                            const Format* texel_format) {
+  std::vector<Value> scaled_texel(texel.size());
```
Comment (dj2) [inline src/verifier.cc]: Why not update in place, why create a new vector?

File: src/verifier.cc [hunk 12]
```diff
@@ -177,18 +188,204 @@ void SetupToleranceForTexels(const ProbeCommand* command,
...
 Verifier::Verifier() = default;
 
 Verifier::~Verifier() = default;
 
+bool Verifier::IsTexelEqualToExpected(const std::vector<Value>& texel,
+                                      const Format* texel_format,
+                                      const ProbeCommand* command,
+                                      const double* tolerance,
+                                      const bool* is_tolerance_percent) {
+  TexelErrorInfo texel_error_info = {};
```
Comment (dj2) [inline src/verifier.cc]: TexelErrorInfo texel_error_info = TexelErrorInfo();  to make CTS happy

File: src/verifier.cc [hunk 13]
```diff
@@ -268,22 +453,30 @@ Result Verifier::Probe(const ProbeCommand* command,
...
+        (command->IsRGBA() ? ", " +
+                                 std::to_string(command->GetA() *
+                                                scale_factor_for_error_report) +
                                  "\n  Actual RGBA: "
                            : "\n  Actual RGB: ") +
-        std::to_string(static_cast<int>(p[0])) + ", " +
-        std::to_string(static_cast<int>(p[1])) + ", " +
-        std::to_string(static_cast<int>(p[2])) +
-        (command->IsRGBA() ? ", " + std::to_string(static_cast<int>(p[3]))
+        std::to_string(actual_color_texel_.r) + ", " +
```
Comment (dj2) [inline src/verifier.cc]: Why do we need actual_color_texel_ can't we just copy the failing vector of values in the loop above? No need for the class value.

File: src/verifier.cc [hunk 14]
```diff
@@ -268,22 +453,30 @@ Result Verifier::Probe(const ProbeCommand* command,
...
 
   if (count_of_invalid_pixels) {
-    const uint8_t* p = ptr + row_stride * (first_invalid_j + y) +
-                       texel_stride * (x + first_invalid_i);
+    const auto& component = texel_format->GetComponents().back();
+    float scale_factor_for_error_report = 1.0f;
+    if (component.mode == FormatMode::kUNorm ||
+        component.mode == FormatMode::kSNorm ||
+        component.mode == FormatMode::kSRGB) {
+      scale_factor_for_error_report = 255.0f;
```
Comment (dj2) [inline src/verifier.cc]: We have the scale factor  for error report here and above why do we need it twice?

File: src/vulkan/format_data.cc [hunk 15]
```diff
@@ -285,15 +285,157 @@ VkFormat ToVkFormat(FormatType type) {
...
+    case VK_FORMAT_R5G5B5A1_UNORM_PACK16:
+    case VK_FORMAT_R5G6B5_UNORM_PACK16:
+    case VK_FORMAT_R8G8_SINT:
+    case VK_FORMAT_R8G8_SNORM:
+    case VK_FORMAT_R8G8_SRGB:
+    case VK_FORMAT_R8G8_SSCALED:
+    case VK_FORMAT_R8G8_UINT:
+    case VK_FORMAT_R8G8_UNORM:
+    case VK_FORMAT_R8G8_USCALED:
+      return 2U;
```
Comment (dj2) [inline src/vulkan/format_data.cc]: Can we sort these in increasing order of size, so the 2U's then the 4Us, etc.

File: src/verifier.cc [hunk 16]
```diff
@@ -177,18 +188,204 @@ void SetupToleranceForTexels(const ProbeCommand* command,
...
+
+  return actual_values;
+}
+
+// If component mode of |texel_format| is FormatMode::kUNorm or
+// ::kSNorm or ::kSRGB, scale the corresponding value in |texel|.
+// Note that we do not scale values with FormatMode::kUInt, ::kSInt,
+// ::kUFloat, ::kSFloat.
+std::vector<Value> ScaleTexelValuesIfNeeded(const std::vector<Value>& texel,
+                                            const Format* texel_format) {
```
Comment (dj2) [inline src/verifier.cc]: Let's rename texel_format to framebuffer_format to make it clear what the format is.

File: src/verifier.cc (hunk 10 above)
Comment (jaebaek) [reply src/verifier.cc]: We had offline discussion and determined to keep it.

File: src/verifier.cc (hunk 16 above)
Comment (jaebaek) [reply src/verifier.cc]: done

File: src/verifier.cc (hunk 11 above)
Comment (jaebaek) [reply src/verifier.cc]: done

File: src/verifier.cc (hunk 14 above)
Comment (jaebaek) [reply src/verifier.cc]: Removed the above one.

File: src/verifier.cc (hunk 13 above)
Comment (jaebaek) [reply src/verifier.cc]: done

File: src/vulkan/format_data.cc (hunk 15 above)
Comment (jaebaek) [reply src/vulkan/format_data.cc]: done

File: src/verifier_test.cc [hunk 17]
```diff
@@ -250,7 +250,7 @@ TEST_F(VerifierTest, ProbeFrameBufferUInt32) {
                             4 * static_cast<uint32_t>(sizeof(uint32_t)), 1, 1,
                             static_cast<const void*>(&frame_buffer));
   EXPECT_TRUE(r.IsSuccess());
-  EXPECT_STREQ("", r.Error());
+  EXPECT_STREQ("", r.Error().c_str());
```
Comment (dj2) [inline src/verifier_test.cc]: This shouldn't be needed. Use:

```
EXPECT_EQ("", r.Error());
```
and it should do string comparisons.

Comment (jaebaek) [reply src/verifier_test.cc]: Done

**Pull request merged**